modules/margin-protocol/src/lib.rs

```diff
@@ -189,6 +217,66 @@ impl<T: Trait> Module<T> {
 		// TODO: implementation
 		unimplemented!()
 	}
+
+	fn _trader_margin_call(who: &T::AccountId) -> DispatchResult {
+		if !<MarginCalledTraders<T>>::contains_key(who) {
+			let threshold = TraderRiskThreshold::get();
+			let margin_level = Self::_margin_level(who, None)?;
+
+			if margin_level <= threshold.margin_call.into() {
+				<MarginCalledTraders<T>>::insert(who, ());
+			} else {
+				return Err(Error::<T>::CannotTraderMarginCall.into());
+			}
+		}
+		Ok(())
+	}
+
+	fn _trader_become_safe(who: &T::AccountId) -> DispatchResult {
+		if <MarginCalledTraders<T>>::contains_key(who) {
+			let threshold = TraderRiskThreshold::get();
+			let margin_level = Self::_margin_level(who, None)?;
+
+			if margin_level > threshold.margin_call.into() {
+				<MarginCalledTraders<T>>::remove(who);
+			} else {
+				return Err(Error::<T>::CannotTraderBecomeSafe.into());
+			}
+		}
+		Ok(())
+	}
+
+	fn _trader_liquidate(who: &T::AccountId) -> DispatchResult {
+		let threshold = TraderRiskThreshold::get();
+		let margin_level = Self::_margin_level(who, None)?;
+
+		if margin_level <= threshold.stop_out.into() {
+			// Close position as much as possible
+			// TODO: print error log
+			<PositionsByTrader<T>>::iter_prefix(who).for_each(|user_position_ids| {
```

Comment:
As closing position will not change margin level, all trader's positions should be closed here.

Plus there should be no `min_price` in `_close_position` on liquidation, and positions should be closed what ever the market bid price is. You could make the `min` parameter as an `Option`.

Reply:
Maybe closed position will change margin level,  because leveraged_held will reduce.

Reply:
> Plus there should be no `min_price` in `_close_position` on liquidation, and positions should be closed what ever the market bid price is. You could make the `min` parameter as an `Option`.
Do I need to use `min_price` in `_bid_price`?
Such as `Price::min_value()`

Reply:
> Maybe closed position will change margin level, because leveraged_held will reduce.

But user's balance will also change. The only actual difference is whether trader's profit/loss and accumulated fees got realized or not.

Reply:
> > Plus there should be no `min_price` in `_close_position` on liquidation, and positions should be closed what ever the market bid price is. You could make the `min` parameter as an `Option`.
> 
> Do I need to use `min_price` in `_bid_price`?
> Such as `Price::min_value()`

Getting is not necessary, just call `_close_position` with `price` as `None`(after refactoring `price` parameter to `Option<Price>` type).